security/src/main/java/io/micronaut/security/token/propagation/TokenPropagationHttpClientFilter.java

```diff
@@ -92,11 +112,25 @@ public TokenPropagationHttpClientFilter(TokenWriter tokenWriter,
         if (token.isPresent()) {
             Object obj = token.get();
             if (obj instanceof String) {
+                boolean targetRequestAlreadyHasToken = hasExistingToken(targetRequest);
                 String tokenValue = (String) obj;
-                tokenWriter.writeToken(targetRequest, tokenValue);
-                return chain.proceed(targetRequest);
+                if (targetRequestAlreadyHasToken) {
+                    if (tokenPropagationConfiguration != null && tokenPropagationConfiguration.isOverwrite()) {
+                        tokenWriter.writeToken(targetRequest, tokenValue);
+                    }
+                } else {
+                    tokenWriter.writeToken(targetRequest, tokenValue);
+                }
             }
         }
         return chain.proceed(targetRequest);
     }
+
+    private boolean hasExistingToken(MutableHttpRequest<?> targetRequest) {
+        String headerName = HttpHeaders.AUTHORIZATION;
+        if (httpHeaderTokenWriterConfiguration != null) {
+            headerName = httpHeaderTokenWriterConfiguration.getHeaderName();
+        }
+        return targetRequest.getHeaders().get(headerName) != null;
+    }
```

Comment:
This assumes the token will be written via headers. It should be the responsibility of the token writer to determine if the token already exists.

Reply:
Good point, but actually now you mention, it's `TokenReader` the one that should determine whether the token exists